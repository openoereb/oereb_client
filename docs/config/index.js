var path = require('canonical-path');

var Package = require('dgeni').Package;

module.exports = new Package('oereb_client_doc', [
  require('dgeni-packages/ngdoc'),
  require('dgeni-packages/nunjucks')
])

  .processor(require('./processors/index-page'))
  .processor(require('./processors/version'))
  .processor(require('./processors/api-data'))
  .processor(require('./processors/guide-data'))

  .config(function (log, readFilesProcessor, writeFilesProcessor) {

    // Set the log level to 'info', switch to 'debug' when troubleshooting
    log.level = 'info';

    // Specify the base path used when resolving relative paths to source and output files
    readFilesProcessor.basePath = path.resolve(__dirname, '../..');

    // Specify our source files that we want to extract
    readFilesProcessor.sourceFiles = [
      { // All of our application files
        include: 'oereb_client/static/js/**/*.js',
        basePath: 'oereb_client/static/js'
      },
      { // Our static Markdown documents
        include: 'docs/content/**/*.md',
        basePath: 'docs/content',
        fileReader: 'ngdocFileReader'
      }
    ];

    // Use the writeFilesProcessor to specify the output folder for the extracted files
    writeFilesProcessor.outputFolder = 'docs/build';

  })

  .config(function (parseTagsProcessor, getInjectables) {
    parseTagsProcessor.tagDefinitions =
      parseTagsProcessor.tagDefinitions.concat(getInjectables(require('./tag-defs')));
  })

  .config(function (templateFinder) {

    // Specify where the templates are located
    templateFinder.templateFolders.unshift(path.resolve(__dirname, 'templates'));

  })

  .config(function (computePathsProcessor, computeIdsProcessor) {

    computePathsProcessor.pathTemplates.push({
      docTypes: ['module'],
      pathTemplate: '${area}/${name}',
      outputPathTemplate: 'partials/${area}/${name}.html'
    });

    computePathsProcessor.pathTemplates.push({
      docTypes: ['componentGroup'],
      pathTemplate: '${area}/${moduleName}/${groupType}',
      outputPathTemplate: 'partials/${area}/${moduleName}/${groupType}.html'
    });

    // create new compute for 'content' type doc
    // indexPage is something new we will be defining later
    computeIdsProcessor.idTemplates.push({
      docTypes: ['content', 'indexPage'],
      getId: function (doc) {
        return doc.fileInfo.baseName;
      },
      getAliases: function (doc) {
        return [doc.id];
      }
    });

    // Build custom paths and set the outputPaths for "content" pages
    computePathsProcessor.pathTemplates.push({
      docTypes: ['content'],
      getPath: function (doc) {
        var docPath = path.dirname(doc.fileInfo.relativePath);
        if (doc.fileInfo.baseName !== 'index') {
          docPath = path.join(docPath, doc.fileInfo.baseName);
        }
        return docPath;
      },
      outputPathTemplate: 'partials/${path}.html'
    });

  })

  .config(function (extractAccessTransform) {
    extractAccessTransform.allowedDocTypes.add('js');
  });
