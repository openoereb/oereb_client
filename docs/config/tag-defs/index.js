module.exports = [
  require('./export'),
  require('./ngInject')
];