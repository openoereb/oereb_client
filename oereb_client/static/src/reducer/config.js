import {createSlice} from "@reduxjs/toolkit";

export const configSlice = createSlice({
  name: 'config',
  initialState: {
    config: {}
  },
  reducers: {
    update: (state, action) => {
      state.config = action.payload;
    }
  }
});

export const {update} = configSlice.actions;

export default configSlice.reducer;
