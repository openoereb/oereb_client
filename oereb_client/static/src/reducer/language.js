import {createSlice} from "@reduxjs/toolkit";

import i18n from "../i18n";

export const languageSlice = createSlice({
  name: 'language',
  initialState: {
    current: null,
    default: null,
    available: []
  },
  reducers: {
    initLanguages: (state, action) => {
      state.default = action.payload.default;
      state.available = action.payload.available;
      const query = new URLSearchParams(window.location.search);
      if (query.has('lang') && state.available.indexOf(query.get('lang')) > -1) {
        state.current = query.get('lang');
      }
      else {
        state.current = state.default;
      }
      i18n.changeLanguage(state.current);
      query.set('lang', state.current);
      window.history.pushState(null, null, '?' + query.toString());
    },
    setLanguage: (state, action) => {
      if (state.available.indexOf(action.payload) > -1) {
        state.current = action.payload;
        i18n.changeLanguage(state.current);
        const query = new URLSearchParams(window.location.search);
        query.set('lang', state.current);
        window.history.pushState(null, null, '?' + query.toString());
      }
    }
  }
});

export const {initLanguages, setLanguage} = languageSlice.actions;

export default languageSlice.reducer;
