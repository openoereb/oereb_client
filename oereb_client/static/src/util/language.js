import {isArray} from "lodash";

export const getLocalizedText = function (multilingualText, language, defaultLanguage) {
  if (isArray(multilingualText) && multilingualText.length > 0) {
    let defaultValue = null;
    for (let i = 0; i < multilingualText.length; i++) {
      if (multilingualText[i]["Language"] === language) {
        return multilingualText[i]["Text"];
      }
      else if (multilingualText[i]["Language"] === defaultLanguage) {
        defaultValue = multilingualText[i]["Text"];
      }
    }
    if (defaultValue) {
      return defaultValue;
    }
    return multilingualText[0]["Text"];
  }
  return null;
};

export const getLocalizedUrl = function (multilingualUrl, language, defaultLanguage) {
  if (isArray(multilingualUrl) && multilingualUrl.length > 0) {
    let defaultValue = null;
    for (let i = 0; i < multilingualUrl.length; i++) {
      if (multilingualUrl[i]["Language"] === language) {
        return multilingualUrl[i]["URL"];
      }
      else if (multilingualUrl[i]["Language"] === defaultLanguage) {
        defaultValue = multilingualUrl[i]["URL"];
      }
    }
    if (defaultValue) {
      return defaultValue;
    }
    return multilingualUrl[0]["URL"];
  }
  return null;
};
