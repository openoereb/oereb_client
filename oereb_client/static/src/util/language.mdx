import {Meta} from '@storybook/addon-docs/blocks';

<Meta title="API Reference/Util/Language" />

# getLocalizedText

Accepts a multilingual array containing localized texts for certain languages,
the requested language and the default language. The localized text of the
requested language is returned as result. If the requested language is not
available in the array, the default language is used as fallback.

```js
getLocalizedText(multilingualText, language, defaultLanguage);
```

### Parameters

| Param | Type | Description |
| --- | --- | --- |
| multilingualText | <code>Array</code> | The array containing the localized texts. |
| language | <code>String</code> | The requested language. |
| defaultLanguage | <code>String</code> | The default language used as fallback. |

### Returns

<code>String</code> The localized text.

# getLocalizedUrl

Accepts a multilingual array containing localized URLs for certain languages,
the requested language and the default language. The localized URL of the
requested language is returned as result. If the requested language is not
available in the array, the default language is used as fallback.

```js
getLocalizedUrl(multilingualUrl, language, defaultLanguage);
```

### Parameters

| Param | Type | Description |
| --- | --- | --- |
| multilingualUrl | <code>Array</code> | The array containing the localized URLs. |
| language | <code>String</code> | The requested language. |
| defaultLanguage | <code>String</code> | The default language used as fallback. |

### Returns

<code>String</code> The localized URL.
