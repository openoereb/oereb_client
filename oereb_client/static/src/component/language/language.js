import PropTypes from 'prop-types';
import React from "react";
import {withNamespaces} from 'react-i18next';
import {useDispatch, useSelector} from "react-redux";

import {setLanguage} from "../../reducer/language";

const OerebLanguage = function({t}) {
  const dispatch = useDispatch();
  const language = useSelector((state) => state.language);
  const currentLanguage = language.current;
  const availableLanguages = language.available;

  const selectLanguage = function(lang) {
    dispatch(setLanguage(lang));
  };

  const languages = availableLanguages.map((language, key) =>
    <li key={key}>
      <button className="dropdown-item" onClick={selectLanguage.bind(this, language)}>
        {language.toUpperCase()}
      </button>
    </li>
  );

  return (
    <div>
      <button className="btn btn-outline-secondary dropdown-toggle"
        type="button"
        data-bs-toggle="dropdown"
        aria-expanded="false"
        title={t('menu.language.title')}>
        {currentLanguage.toUpperCase()}
      </button>
      <ul className="dropdown-menu">
        {languages}
      </ul>
    </div>
  );
};

OerebLanguage.propTypes = {
  t: PropTypes.func.isRequired
};

export default withNamespaces()(OerebLanguage);
