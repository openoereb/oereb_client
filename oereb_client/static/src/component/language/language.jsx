import React from "react";
import {useTranslation} from "react-i18next";
import {useDispatch, useSelector} from "react-redux";

import {hideExtract, loadExtract, showError, showExtract} from "../../reducer/extract";
import {updateHistory} from "../../reducer/history";
import {setLanguage} from "../../reducer/language";
import {queryExtractById} from "../../request/extract";
import {TooManyRequestsError} from "../../util/error";
import {showWarning} from "../../reducer/message";

/**
 * This component provides the language selector.
 */
const OerebLanguage = function() {
  const {t} = useTranslation();
  const dispatch = useDispatch();
  const config = useSelector((state) => state.config).config;
  const serviceUrl = config.service_url;
  const language = useSelector((state) => state.language);
  const currentLanguage = language.current;
  const availableLanguages = language.available;
  const extract = useSelector((state) => state.extract);

  const selectLanguage = function(lang) {
    if (lang !== currentLanguage) {
      dispatch(setLanguage(lang));
      if (extract.visible) {
        dispatch(loadExtract({
          egrid: extract.egrid,
          identdn: extract.identdn,
          number: extract.number,
          zoom: false
        }));
        queryExtractById(
          serviceUrl,
          extract.egrid,
          extract.identdn,
          extract.number,
          config.extract_json_timeout,
          lang
        )
          .then((data) => {
            dispatch(showExtract(data));
            dispatch(updateHistory(data));
          })
          .catch((error) => {
            if (error instanceof TooManyRequestsError) {
              dispatch(hideExtract());
              dispatch(showWarning(t('extract.error.too_many_requests'), true));
            }
            else {
              dispatch(showError());
            }
          });
      }
    }
  };

  const languages = availableLanguages.map((language, key) =>
    <li key={key}>
      <button className="dropdown-item" onClick={selectLanguage.bind(this, language)}>
        {language.toUpperCase()}
      </button>
    </li>
  );

  return (
    <div className="oereb-client-language-selector">
      <button className="btn btn-outline-secondary dropdown-toggle"
        type="button"
        data-bs-toggle="dropdown"
        aria-expanded="false"
        title={t("menu.language.title")}>
        {currentLanguage.toUpperCase()}
      </button>
      <ul className="dropdown-menu">
        {languages}
      </ul>
    </div>
  );
};

export default OerebLanguage;
