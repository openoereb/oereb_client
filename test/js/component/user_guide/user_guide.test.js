import React from "react";
import {render} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import {act} from "react-dom/test-utils";
import {Provider} from "react-redux";

import OerebUserGuide from "../../../../oereb_client/static/src/component/user_guide/user_guide";
import {update} from "../../../../oereb_client/static/src/reducer/config";
import {initLanguages} from "../../../../oereb_client/static/src/reducer/language";
import MainStore from "../../../../oereb_client/static/src/store/main";
import { async } from "regenerator-runtime";

describe('user guide component', () => {

  let component;
  let user;

  beforeEach(() => {
    act(() => {
      MainStore.dispatch(initLanguages({
        default: 'de',
        available: ['de']
      }));
    });
    act(() => {
      MainStore.dispatch(update({
        user_guide: 'https://example.com/guide/{lang}.pdf'
      }));
    });
    component = render(
      <Provider store={MainStore}>
        <OerebUserGuide />
      </Provider>
    );
    user = userEvent.setup();
    window.open = jest.fn();
  });

  afterEach(() => {
    window.open.mockClear();
  });

  it('should render button', () => {
    expect(component.asFragment()).toMatchSnapshot();
  });

  it('should not render button', () => {
    act(() => {
      MainStore.dispatch(update({}));
    });
    const wrapper = render(
      <Provider store={MainStore}>
        <OerebUserGuide />
      </Provider>
    );
    expect(wrapper.asFragment()).toMatchSnapshot();
  });

  it('should call window.open', async () => {
    await user.click(component.container.querySelector('button'));
    expect(window.open.mock.calls).toHaveLength(1);
    expect(window.open.mock.calls[0][0]).toEqual('https://example.com/guide/de.pdf');
    expect(window.open.mock.calls[0][1]).toEqual('_blank');
    window.open.mockReset();
  });

});