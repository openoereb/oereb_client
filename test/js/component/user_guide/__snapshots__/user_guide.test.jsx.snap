// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`user guide component > should not render button 1`] = `<DocumentFragment />`;

exports[`user guide component > should render button 1`] = `
<DocumentFragment>
  <button
    class="btn btn-outline-secondary oereb-client-user-guide"
    title="menu.user_guide.title"
    type="button"
  >
    <i
      class="bi bi-question-circle"
    />
  </button>
</DocumentFragment>
`;
